sw/device/lib/dif/dif_i2c.c

```diff
@@ -0,0 +1,681 @@
+      break;
+    case kDifI2cLevel4Byte:
+      rx_level_value = I2C_FIFO_CTRL_RXILVL_RXLVL4;
+      break;
+    case kDifI2cLevel8Byte:
+      rx_level_value = I2C_FIFO_CTRL_RXILVL_RXLVL8;
+      break;
+    case kDifI2cLevel16Byte:
+      rx_level_value = I2C_FIFO_CTRL_RXILVL_RXLVL16;
+      break;
+    case kDifI2cLevel30Byte:
+      rx_level_value = I2C_FIFO_CTRL_RXILVL_RXLVL30;
+      break;
+    default:
+      return kDifI2cResultBadArg;
+  }
+
+  ptrdiff_t fmt_level_value;
+  switch (fmt_level) {
+    case kDifI2cLevel1Byte:
+      fmt_level_value = I2C_FIFO_CTRL_FMTILVL_FMTLVL1;
+      break;
+    case kDifI2cLevel4Byte:
+      fmt_level_value = I2C_FIFO_CTRL_FMTILVL_FMTLVL4;
+      break;
+    case kDifI2cLevel8Byte:
+      fmt_level_value = I2C_FIFO_CTRL_FMTILVL_FMTLVL8;
+      break;
+    case kDifI2cLevel16Byte:
+      fmt_level_value = I2C_FIFO_CTRL_FMTILVL_FMTLVL16;
+      break;
+    default:
+      return kDifI2cResultBadArg;
+  }
+
+  uint32_t ctrl_value =
+      mmio_region_read32(i2c->base_addr, I2C_FIFO_CTRL_REG_OFFSET);
+  ctrl_value = bitfield_set_field32(
+      ctrl_value, (bitfield_field32_t){.mask = I2C_FIFO_CTRL_RXILVL_MASK,
+                                       .index = I2C_FIFO_CTRL_RXILVL_OFFSET,
+                                       .value = rx_level_value});
+  ctrl_value = bitfield_set_field32(ctrl_value,
+                                    (bitfield_field32_t){
+                                        .mask = I2C_FIFO_CTRL_FMTILVL_MASK,
+                                        .index = I2C_FIFO_CTRL_FMTILVL_OFFSET,
+                                        .value = fmt_level_value,
+                                    });
+  mmio_region_write32(i2c->base_addr, I2C_FIFO_CTRL_REG_OFFSET, ctrl_value);
+
+  return kDifI2cResultOk;
+}
+
+dif_i2c_result_t dif_i2c_irq_get(const dif_i2c_t *i2c, dif_i2c_irq_type_t type,
+                                 bool *flag_out) {
+  if (i2c == NULL || flag_out == NULL) {
+    return kDifI2cResultBadArg;
+  }
+
+  uint32_t index;
+  switch (type) {
+    case kDifI2cIrqTypeFmtWatermarkUnderflow:
+      index = I2C_INTR_STATE_FMT_WATERMARK;
+      break;
+    case kDifI2cIrqTypeRxWatermarkOverflow:
+      index = I2C_INTR_STATE_RX_WATERMARK;
+      break;
+    case kDifI2cIrqTypeFmtFifoOverflow:
+      index = I2C_INTR_STATE_FMT_OVERFLOW;
+      break;
+    case kDifI2cIrqTypeRxFifoOverflow:
+      index = I2C_INTR_STATE_RX_OVERFLOW;
+      break;
+    case kDifI2cIrqTypeNak:
+      index = I2C_INTR_STATE_NAK;
+      break;
+    case kDifI2cIrqTypeSclInterference:
+      index = I2C_INTR_STATE_SCL_INTERFERENCE;
+      break;
+    case kDifI2cIrqTypeSdaInterference:
+      index = I2C_INTR_STATE_SDA_INTERFERENCE;
+      break;
+    case kDifI2cIrqTypeClockStretchTimeout:
+      index = I2C_INTR_STATE_STRETCH_TIMEOUT;
+      break;
+    case kDifI2cIrqTypeSdaUnstable:
+      index = I2C_INTR_STATE_SDA_UNSTABLE;
+      break;
+    default:
+      return kDifI2cResultBadArg;
+  }
```

Comment:
Can you rebase and refactor this to a `static` function that uses `I2C_INTR_COMMON_*` and remove the `switch` statements in other functions too?

Reply:
Please see: #2222 and #1881.

Reply:
Done.